pkg/providers/aws/provider_redis_test.go

```diff
@@ -0,0 +1,243 @@
+			}
+		})
+	}
+}
+
+func TestAWSRedisProvider_createRedisCluster(t *testing.T) {
+	scheme := runtime.NewScheme()
+	err := v1.AddToScheme(scheme)
+	err = corev1.AddToScheme(scheme)
+	err = v1alpha1.SchemeBuilder.AddToScheme(scheme)
+	if err != nil {
+		t.Fatal("failed to build scheme", err)
+	}
+
+	sc := &StrategyConfig{
+		Region:      "eu-west-1",
+		RawStrategy: json.RawMessage("{}"),
+	}
+	rawStratCfg, err := json.Marshal(sc)
+	if err != nil {
+		t.Fatal("failed to marshal strategy config", err)
+	}
+
+	cases := []struct {
+		name           string
+		instance       *v1alpha1.Redis
+		client         client.Client
+		configMgr      *ConfigManagerMock
+		credentialMgr  *CredentialManagerMock
+		expectedError  error
+		expectedResult *providers.RedisCluster
+	}{
+		{
+			name: "test redis cluster is created as expected",
+			client: fake.NewFakeClientWithScheme(scheme, &v1alpha1.Redis{
+				ObjectMeta: controllerruntime.ObjectMeta{
+					Name:      "test",
+					Namespace: "test",
+				},
+			}, &corev1.ConfigMap{
+				ObjectMeta: controllerruntime.ObjectMeta{
+					Name:      "cloud-resources-aws-strategies",
+					Namespace: "kube-system",
+				},
+				Data: map[string]string{
+					"redis": fmt.Sprintf("{\"test\": %s}", string(rawStratCfg)),
+				},
+			}),
+			instance: &v1alpha1.Redis{
+				ObjectMeta: controllerruntime.ObjectMeta{
+					Name:      "test",
+					Namespace: "test",
+				},
+			},
+			configMgr: &ConfigManagerMock{
+				ReadStorageStrategyFunc: func(ctx context.Context, rt providers.ResourceType, tier string) (config *StrategyConfig, e error) {
+					return sc, nil
+				},
+			},
+			expectedError:  nil,
+			expectedResult: nil,
+		},
+	}
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			p := &AWSRedisProvider{
+				Client:            tc.client,
+				CredentialManager: tc.credentialMgr,
+				ConfigManager:     tc.configMgr,
+			}
+			redisConfig, _, err := p.getRedisConfig(context.TODO(), tc.instance)
+			if err != nil {
+				t.Fatal("", err)
+			}
+
+			mockSvc := &mockElasticacheClient{}
+			redis, err := createRedisCluster(mockSvc, redisConfig)
+			if redis != tc.expectedResult {
+				t.Fatalf("unexpected outcome, expected %s but got %s", tc.expectedResult, redis)
+			}
+			if err != tc.expectedError {
+				t.Fatalf("unexpected error, expected %s but got %s", tc.expectedError, redis)
+			}
+		})
+	}
+}
+
+func TestAWSRedisProvider_deleteRedisCluster(t *testing.T) {
+	scheme := runtime.NewScheme()
+	err := v1.AddToScheme(scheme)
+	err = corev1.AddToScheme(scheme)
+	err = v1alpha1.SchemeBuilder.AddToScheme(scheme)
+	if err != nil {
+		t.Fatal("failed to build scheme", err)
+	}
+
+	sc := &StrategyConfig{
+		Region:      "eu-west-1",
+		RawStrategy: json.RawMessage("{}"),
+	}
+	rawStratCfg, err := json.Marshal(sc)
+	if err != nil {
+		t.Fatal("failed to marshal strategy config", err)
+	}
+
+	cases := []struct {
+		name          string
```

Comment:
Do you find breaking it up this way means quite a log of duplication? Would it cause less duplication to have create delete etc as cases under a test_redis method?

Reply:
Sure, we can look into refactoring that

Reply:
Left it separated for now because we're testing different functions